stan/math/prim/arr/functor/map_rect_mpi.hpp

```diff
@@ -0,0 +1,147 @@
+        if(distributed_data::get_const_instance().find(uid) == distributed_data::get_const_instance().cend()) {
+          //std::cout << "Job " << R << " registers data..." << std::endl;
+        } else {
+          //std::cout << "UID = " << uid << " is ALREADY distributed." << std::endl;
+          return;
+        }
+
+        std::vector<int> meta(4, 0);
+        
+        if(R == 0) {
+          // initiate on the root call of this function on the workers
+          mpi_broadcast_command<stan::math::mpi_distributed_apply<distributed_map_rect_data> >();
+
+          meta[0] = uid;
+          meta[1] = x_r.size();
+          meta[2] = x_r[0].size();
+          meta[3] = x_i[0].size();
+        }
+
+        boost::mpi::broadcast(world, meta.data(), 4, 0);
+
+        distributed_map_rect_data data(meta[0]);
+
+        const std::size_t N = meta[1];
+        const std::size_t X_r = meta[2];
+        const std::size_t X_i = meta[3];
+
+        data.N_ = N;
+
+        //std::cout << "worker " << R << " / " << W << " registers shapes " << N << ", " << X_r << ", " << X_i << std::endl;
+
+        const std::vector<int> chunks = mpi_map_chunks(N, 1);
+        
+        data.x_r_.resize(chunks[R]);
+        data.x_i_.resize(chunks[R]);
+
+           // flatten data and send out/recieve using scatterv
+        if(X_r > 0) {
+          const std::vector<double> world_x_r = to_array_1d(x_r);
+          const std::vector<int> chunks_x_r = mpi_map_chunks(N, X_r);
+          std::vector<double> flat_x_r_local(chunks_x_r[R]);
+
+          boost::mpi::scatterv(world, world_x_r.data(), chunks_x_r, flat_x_r_local.data(), 0);
+
+          // now register data
+          for(std::size_t i = 0; i != chunks[R]; ++i)
+            data.x_r_[i] = std::vector<double>(flat_x_r_local.begin() + i * X_r, flat_x_r_local.begin() + (i+1) * X_r);
+          
+          //std::cout << "Job " << R << " got " << flat_x_r_local.size() << " real data " << std::endl;
+        }
+        if(X_i > 0) {
+          const std::vector<int> world_x_i = to_array_1d(x_i);
+          const std::vector<int> chunks_x_i = mpi_map_chunks(N, X_i);
+          std::vector<int> flat_x_i_local(chunks_x_i[R]);
+
+          boost::mpi::scatterv(world, world_x_i.data(), chunks_x_i, flat_x_i_local.data(), 0);
+
+          // now register data
+          for(std::size_t i = 0; i != chunks[R]; ++i)
+            data.x_i_[i] = std::vector<int>(flat_x_i_local.begin() + i * X_i, flat_x_i_local.begin() + (i+1) * X_i);
+          
+          //std::cout << "Job " << R << " got " << flat_x_i_local.size() << " int data " << std::endl;
+        }
+
+        distributed_data::get_mutable_instance().insert(std::make_pair(uid, data));
+
+        //std::cout << "Job " << R << " done caching data for uid = " << meta[0] << "." << std::endl;
+        
+        return;
+      }
+    }
+
+    template <typename F>
+    std::vector<double>
+    map_rect_mpi(const std::vector<double>& eta,
```

Comment:
This one also shouldn't parallelize?

Reply:
It should... but I would be fine to make it non-parallel for the very first pull. The thing is, I wanted this stuff to work in an actual Stan program; until I figure out the right MPI code I am very reluctant to copy&paste the code which is right now in rev. Once we think the rev code is good to copy (and simplify) we should do that. Even better would be to figure out code which handles all cases at the same time.